nuttx-patches/priority_restoration_fix.patch

```diff
@@ -0,0 +1,448 @@
+  *
+  * Description:
+@@ -568,7 +650,7 @@ static int sem_restoreholderprioA(FAR struct semholder_s *pholder,
+   FAR struct tcb_s *rtcb = this_task();
+   if (pholder->htcb != rtcb)
+     {
+-      return sem_restoreholderprio(pholder, sem, arg);
++      return sem_restoreholderprio(pholder->htcb, sem, arg);
+     }
+ 
+   return 0;
+@@ -586,9 +668,22 @@ static int sem_restoreholderprioB(FAR struct semholder_s *pholder,
+                                   FAR sem_t *sem, FAR void *arg)
+ {
+   FAR struct tcb_s *rtcb = this_task();
++
+   if (pholder->htcb == rtcb)
+     {
+-      (void)sem_restoreholderprio(pholder, sem, arg);
++
++   /* The running task has given up a count on the semaphore */
++
++#if CONFIG_SEM_PREALLOCHOLDERS == 0
++      /* In the case where there are only 2 holders. This step
++       * is necessary to insure we have space. Release the holder
++       * if all counts have been given up. before reprioritizing
++       * causes a context switch.
++       */
++
++       sem_findandfreeholder(sem, rtcb);
++#endif
++      (void)sem_restoreholderprio(rtcb, sem, arg);
+       return 1;
+     }
+ 
+@@ -637,11 +732,11 @@ static inline void sem_restorebaseprio_irq(FAR struct tcb_s *stcb,
+    * next highest pending priority.
+    */
+ 
+-  if (stcb)
++  if (stcb != NULL)
+     {
+       /* Drop the priority of all holder threads */
+ 
+-      (void)sem_foreachholder(sem, sem_restoreholderprio, stcb);
++      (void)sem_foreachholder(sem, sem_restoreholderprioall, stcb);
+     }
+ 
+   /* If there are no tasks waiting for available counts, then all holders
+@@ -701,7 +796,7 @@ static inline void sem_restorebaseprio_task(FAR struct tcb_s *stcb,
+    * next highest pending priority.
+    */
+ 
+-  if (stcb)
++  if (stcb != NULL)
+     {
+       /* The currently executed thread should be the lower priority
+        * thread that just posted the count and caused this action.
+@@ -735,18 +830,8 @@ static inline void sem_restorebaseprio_task(FAR struct tcb_s *stcb,
+    * counts, then we need to remove it from the list of holders.
+    */
+ 
+-  pholder = sem_findholder(sem, rtcb);
+-  if (pholder)
+-    {
+-      /* When no more counts are held, remove the holder from the list.  The
+-       * count was decremented in sem_releaseholder.
+-       */
++  sem_findandfreeholder(sem, rtcb);
+ 
+-      if (pholder->counts <= 0)
+-        {
+-          sem_freeholder(sem, pholder);
+-        }
+-    }
+ }
+ 
+ /****************************************************************************
+@@ -817,18 +902,21 @@ void sem_destroyholder(FAR sem_t *sem)
+    */
+ 
+ #if CONFIG_SEM_PREALLOCHOLDERS > 0
+-  if (sem->hhead)
++  if (sem->hhead != NULL)
+     {
+       serr("ERROR: Semaphore destroyed with holders\n");
++      DEBUGASSERT(sem->hhead != NULL);
+       (void)sem_foreachholder(sem, sem_recoverholders, NULL);
+     }
+ #else
+-  if (sem->holder.htcb)
++  if (sem->holder[0].htcb != NULL || sem->holder[0].htcb != NULL)
```

Comment:
@bkueng - Thank you

Yes there were a few more backward DEBUGASSERT 

it will go in upstream with other changes then comeback herein.